src/core/broker_config.cpp

```diff
@@ -54,89 +74,171 @@ bool ValidateAndStoreCid(const json& val, BrokerConfig& config)
-          return true;
-        }
-      }
-      else if (type == "dynamic" && !val.contains("device_id"))
-      {
-        config.settings.SetDynamicUid(static_cast<uint16_t>(manufacturer_id));
-        return true;
-      }
+      LogParseError(log, "When \"/uid/type\" is \"static\", the \"uid\" object must contain a \"device_id\" field.");
+      return false;
     }
+    if (!val["device_id"].is_number_integer())
+    {
+      LogParseError(log, "The value for setting \"/uid/device_id\" was of invalid type \"%s\"",
+                    val["device_id"].type_name());
+      return false;
+    }
+
+    const int64_t device_id = val["device_id"];
+    if (device_id < 0 || device_id > 0xffffffff)
+    {
+      LogParseError(log, "Static UID: \"%" PRId64 "\" is not a valid Device ID.", device_id);
+      return false;
+    }
+
+    RdmUid static_uid;
+    static_uid.manu = static_cast<uint16_t>(manufacturer_id);
+    static_uid.id = static_cast<uint32_t>(device_id);
+    config.settings.SetStaticUid(static_uid);
+    return true;
+  }
+  else if (type == "dynamic")
+  {
+    if (val.contains("device_id"))
+    {
+      LogParseError(log,
+                    "When \"/uid/type\" is \"dynamic\", the \"uid\" object must not contain a \"device_id\" field.");
+      return false;
+    }
+    config.settings.SetDynamicUid(static_cast<uint16_t>(manufacturer_id));
+    return true;
+  }
+  else
+  {
+    LogParseError(log, "The field \"/uid/type\" must be one of \"static\" or \"dynamic\".");
+    return false;
   }
-  return false;
 }
 
 // Store a generic string.
-bool ValidateAndStoreString(const json& val, std::string& string, size_t max_size, bool truncation_allowed = true)
+bool ValidateAndStoreString(const char* key_ptr, const json& val, std::string& string, size_t max_size,
+                            rdmnet::BrokerLog* log, bool truncation_allowed = true)
 {
   const std::string str_val = val;
-  if (!str_val.empty())
+  if (str_val.empty())
+  {
+    LogParseError(log, "Empty string is not allowed for field \"%s\"", key_ptr);
+    return false;
+  }
+
+  if (str_val.length() > max_size)
   {
     if (truncation_allowed)
     {
       string = str_val.substr(0, max_size);
+      if (log)
+      {
+        log->Notice("Configuration file: Truncating overlong string \"%s\" for field \"%s\" to \"%s\".",
+                    str_val.c_str(), key_ptr, string.c_str());
+      }
       return true;
     }
-    else if (str_val.length() <= max_size)
+    else
     {
-      string = str_val;
-      return true;
+      LogParseError(log, "String value \"%s\" is too long for field \"%s\" of maximum length %zu.", str_val.c_str(),
+                    key_ptr, max_size);
+      return false;
     }
   }
-  return false;
+
+  string = str_val;
+  return true;
 }
 
+// Get the printf-style format string of a given integral type
+template <typename IntType>
+class FormatStringOf
+{
+public:
+  static const char* value;
```

Comment:
I doubt it will be misused given how it is setup, but should this be inititalized to nullptr for safety? Or is that not possible because of specialization?

Reply:
I am not able to do that without getting a compile error. However, if you try to use `FormatStringOf<some_unspecialized_type>`, e.g. some type that hasn't been specialized on the lines below, you will get a link error.